---
output: 
  html_document:
    fig_crop: no
title: "PCGR output: **`r GOI`**"
date: "`r paste(date(),Sys.timezone())`"
---

#  `r GOI` metadata 
Gene symbol: **`r GOI`**  
UniProtKB ID: **`r GOI_uniprot_id`**  
Gene product name: **`r GOI_protein_feature_annotation$LABEL[GOI_protein_feature_annotation$TYPE == "CHAIN"]`**  
Chromosome: **`r GOI_CHR`**  
Strand: **`r GOI_exon_annotation_main_transcript$strand[1]`**  
Number of unique Ensembl transcripts: **`r length(unique(GOI_exon_annotation$ensembl_transcript_id))`**  
Number of unique Ensembl exons for all transcripts: **`r length(unique(GOI_exon_annotation$ensembl_exon_id))`**  

## Primary representative transcript:
Ensembl gene ID **`r GOI_ENSG`**  
Ensembl transcript ID **`r GOI_TRANSCRIPT`**  
Number of exons: **`r length(unique(GOI_exon_annotation_main_transcript$ensembl_exon_id))`**  
Transcript length: **`r GOI_exon_annotation_main_transcript$transcript_length[1]`**  
Coding sequence length: **`r GOI_exon_annotation_main_transcript$cds_length[1]`**  
Transcript start position: **`r GOI_exon_annotation_main_transcript$transcript_start[1]`**  
Transcript end position: **`r GOI_exon_annotation_main_transcript$transcript_end[1]`**  

# cBioPortal Query Metadata

Total number of studies returned by cBioPortal: **`r length(all.studies[,1])`**  
Total number of unique case IDs returned by cBioPortal: **`r (length(unique(master_case_df$case_id)))`**  

Total number of studies with mutation/variant data available:  **`r length(all.mut.studies.filtered)`**  
Total number of cases queried for **`r GOI`** variants: **`r sum(!duplicated(all_mut_cases$altered_case_id))`**  
Total number of studies with **`r GOI`** variants: **`r length(unique(all_mut_cases$study))`**  
Total number of unique cases represented in filtered mutation data: **`r length(unique(GOI_cBP_mutations$altered_case_id))`**  
Total number of unique variants for **`r GOI`** returned:  **`r length(unique(GOI_cBP_mutations$amino_acid_change))`**  

```{r echo=FALSE}
kable(multi_mut_table, row.names = FALSE, align = 'l',caption = "Cases with multiple variants")
```

```{r echo=FALSE}
kable(all_tissue_types_table, row.names = FALSE, align = 'l',caption = "Tissue types in all cases") #fix duplicates
```

# Protein features: Domains, Regions, Motifs and Modifications

```{r echo=FALSE, fig.height=5, fig.width=10, message=FALSE, warning=FALSE, out.width='\\textwidth', strip.white=TRUE}
F1_magick_img <- image_graph(width = 1500, height = 1000,res = 150, clip = FALSE) #Graphics device that produces a Magick image
lolliplot(SNP.gr = F1_variants,
          features = F1_features,
          ranges = F1_ranges,
          ylab = FALSE,
          xaxis = F1_x_axis)
invisible(dev.off()) # shuts down graphics device
par(mar=c(0,0,0,0)) # (bottom,left,top,right)
plot(image_trim(F1_magick_img))
```


```{r echo=FALSE}
kable(GOI_protein_feature_annotation, row.names = FALSE, caption = "All features (minus variants and sequence conflicts)")
```


# Filtered Ensembl Transcripts (Union of Exonic sequence)

All transcripts are ploted to their relative genomic positions with all shared intronic regions removed. Transcription start sites are always on the left regardless of coding strand. **'Main' representative transcript used for variant alignment is in red**  
```{r echo=FALSE,fig.width=14,fig.height=10,out.width='\\textwidth',strip.white=TRUE}
F2_magick_img <- image_graph(width = 2000, height = 2000,res = 150, clip = FALSE)
lolliplot(SNP.gr = F2_variants,
             features = F2_features,
             ylab.gp = gpar(cex=1.25,fontface="bold"),
             ylab = "Filtered transcripts [Union of exonic sequence]\n\n",
             xaxis = F2_x_axis,
             yaxis = FALSE)
invisible(dev.off())
par(mar=c(0,0,0,0)) # (bottom,left,top,right)
plot(image_trim(F2_magick_img))
```

'Main' transcript: **`r GOI_TRANSCRIPT`**  
Transcript support descriptions:  
[Link](https://useast.ensembl.org/info/genome/genebuild/transcript_quality_tags.html)  

```{r echo=FALSE}
kable(GOI_transcript_support[,2:5], row.names = FALSE, caption = "Filtered transcripts returned by ensemble with support annotations")
```

# All cBioPortal variants returned (n = **`r length(F3_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
F3_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
lolliplot(SNP.gr = F3_variants ,
          features = F3_features, 
          ranges = F3_ranges, 
          xaxis = F3_x_axis,
          yaxis.gp = gpar(cex = 1.5),
          xaxis.gp = gpar(cex = 1.5),
          cex = 1,
          ylab = FALSE,
          jitter = "label",
          legend =  legend)
invisible(dev.off())
par(mar=c(0,1.5,0,0)) # (bottom,left,top,right)
plot(image_trim(F3_magick_img))
# yaxis label
grid.text(paste0("Variant count within cBP data"),
          x=0.01,
          y=0.6,
          just="top",
          rot=90,
          gp=gpar(cex=.75, fontface="bold")
          )
```

### cBioPortal Missense mutations (n = **`r length(F4_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F4_variants$score)>0){
  F4_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F4_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F4_types],
                          fill = legend$fill[legend$labels %in% F4_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  par(mar=c(0,1.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F4_magick_img))
  # yaxis label
  grid.text(paste0("Variant count within cBP data"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=.75, fontface="bold")
            )
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Frameshifts, stops gained/lost and start lost (n = **`r length(F5_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F5_variants$score)>0){
  F5_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F5_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F5_types],
                          fill = legend$fill[legend$labels %in% F5_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  par(mar=c(0,1.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F5_magick_img))
  # yaxis label
  grid.text(paste0("Variant count within cBP data"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=.75, fontface="bold")
            )
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Insertions/Deletions  (n = **`r length(F6_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F6_variants$score)>0){
  F6_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F6_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F6_types],
                          fill = legend$fill[legend$labels %in% F6_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  par(mar=c(0,1.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F6_magick_img))
  # yaxis label
  grid.text(paste0("Variant count within cBP data"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=.75, fontface="bold")
            )
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Splice regions and splice sites (n = **`r length(F7_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F7_variants$score)>0){
  F7_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F7_variants ,
            features = c(F7_features,F3_features), 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F7_types],
                          fill = legend$fill[legend$labels %in% F7_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  par(mar=c(0,1.5,1.5,0)) # (bottom,left,top,right)
  plot(image_trim(F7_magick_img))
  # yaxis label
  grid.text(paste0("Variant count within cBP data"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=.75, fontface="bold")
            )
  grid.text(paste0("[ Main transcript exons in grey and red as top feature row ]"),
            x=0.05,
            y=0.95,
            just="left",
            gp=gpar(cex=1, fontface="bold")
            )
} else cat("\tNone returned by query\n\n")
```
```{r echo=FALSE}
kable(F7_exons, row.names = FALSE, caption = "Main transcript exons key", align = "l")
```

```{r echo=FALSE,fig.width=12,fig.height=12,out.width='\\textwidth',strip.white=TRUE}
if(length(F7_variants$score)>0){
  F8_magick_img <- image_graph(width = 2000, height = 2000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F8_variants,
               features = F8_features,
               ylab.gp = gpar(cex=1.25,fontface="bold"),
               ylab = "Filtered transcripts [Union of exonic sequence]",
               xaxis = F2_x_axis,
               yaxis.gp = gpar(cex = 1.5),
               xaxis.gp = gpar(cex = 1.5),
               cex = 1,
               jitter = "label",
               legend = list(labels = legend$labels[legend$labels %in% F7_types],
                            fill = legend$fill[legend$labels %in% F7_types],
                            cex = 1.5
                            )
               )
  invisible(dev.off())
  par(mar=c(0,0,0,0)) # (bottom,left,top,right)
  plot(image_trim(F8_magick_img))
} else cat("\tNone returned by query\n\n")
```


# All gnomAD variants returned (n = **`r length(F9_gnomAD$score)`**)

```{r echo=FALSE, fig.width=14,fig.height=4.5,out.width='\\textwidth'}
F9_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
lolliplot(SNP.gr = F9_gnomAD,
          features = F9_features, 
          ranges = F9_ranges, 
          xaxis = F9_x_axis,
          yaxis.gp = gpar(cex = 1.5),
          xaxis.gp = gpar(cex = 1.5),
          cex = 1,
          ylab = FALSE,
          jitter = "label",
          legend =  legend)
invisible(dev.off())
par(mar=c(0,1.5,0,0)) # (bottom,left,top,right)
plot(image_trim(F9_magick_img))
# yaxis label
grid.text(paste0("log2(allele count) within cBP data [max count = ",max_gnomAD_allele_count,"]"),
          x=0.01,
          y=0.6,
          just="top",
          rot=90,
          gp=gpar(cex=.75, fontface="bold")
          )
```

## gnomAD Missense mutations (n = **`r length(F7_variants$score)`**)

## gnomAD Frameshifts and stops gained/lost (n = **`r length(F5_variants$score)`**)

## gnomAD Insertions/Deletions  (n = **`r length(F6_variants$score)`**)

## gnomAD Splice regions and splice sites (acceptors/donors) (n = **`r length(F7_variants$score)`**)
```{r echo=FALSE, fig.asp=1, fig.width=10,out.width='\\textwidth'}
lolliplot(SNP.gr = list(A1 = GRanges(),A2 = F5_gnomAD_overlap, A3 = F5_cBP_overlap),
          features = list(A3 = Feature_legend, A4 = No_name_features, A5 = No_name_features),
          ranges = F5_ranges, 
          xaxis = FALSE,#,F5_x_axis,F5_x_axis)
          cex = 0.75,
          ylab = c("","\tcBP [# of occurances]\n",paste0("\t\tgnomAD [log2(allele count) ; max = ",max_gnomAD_allele_count,"]\n")),
          jitter = "label",
          legend =  legend)
grid.text(paste0("All variants overlapping between cBioPortal and gnomAD"),
          x=.5,
          y=0.9,
          just="top", 
          gp=gpar(cex=1.5, fontface="bold")
          )
```

```{r echo=FALSE}
kable(overlap_summary_table, row.names = FALSE, caption = "All variants shared between cBP and gnomAD", align = "l")
```

# Gene fusions returned:

```{r echo=FALSE}
kable(GOI_cBP_fusions[,c("case_id","amino_acid_change","genetic_profile_id")], row.names = FALSE)
```
