---
output: 
  html_document:
    fig_crop: no
title: "PCGR output: **`r GOI`**"
date: "`r paste(date(),Sys.timezone())`"
---

<style>
    body .main-container {
        max-width: 1200px;
    }
</style>

##  `r GOI` metadata 

Gene symbol: **`r GOI`**  
UniProtKB ID: **`r GOI_uniprot_id`**  
Gene product name: **`r GOI_protein_feature_annotation$LABEL[GOI_protein_feature_annotation$TYPE == "CHAIN"]`**  
Chromosome: **`r GOI_CHR`**  
Strand: **`r GOI_exon_annotation_main_transcript$strand[1]`**  
Number of unique Ensembl transcripts: **`r length(unique(GOI_exon_annotation$ensembl_transcript_id))`**  
Number of unique Ensembl exons for all transcripts: **`r length(unique(GOI_exon_annotation$ensembl_exon_id))`**  
Genomic start (GRCh38): **`r min(as.numeric(row.names(union_transcripts_relative_pos_key)))`**  
Genomic end (GRCh38): **`r max(as.numeric(row.names(union_transcripts_relative_pos_key)))`**  

## Primary representative transcript:

Ensembl gene ID **`r GOI_ENSG`**  
Ensembl transcript ID **`r GOI_TRANSCRIPT`**  
Number of exons: **`r length(unique(GOI_exon_annotation_main_transcript$ensembl_exon_id))`**  
Transcript length: **`r GOI_exon_annotation_main_transcript$transcript_length[1]`**  
Coding sequence length: **`r GOI_exon_annotation_main_transcript$cds_length[1]`**  
Transcript start position: **`r as.character(GOI_exon_annotation_main_transcript$transcript_start[1])`**  
Transcript end position: **`r as.character(GOI_exon_annotation_main_transcript$transcript_end[1])`**  

## cBioPortal Query Metadata

Total number of studies returned by cBioPortal: **`r length(all.studies[,1])`**  
Total number of unique case IDs returned by cBioPortal: **`r (length(unique(master_case_df$case_id)))`**  

Total number of studies with mutation/variant data available:  **`r length(all.mut.studies.filtered)`**  
Total number of cases queried for **`r GOI`** variants: **`r sum(!duplicated(all_mut_cases$altered_case_id))`**  
Total number of studies with **`r GOI`** variants: **`r length(unique(all_mut_cases$study))`**  
Total number of unique cases represented in filtered mutation data: **`r length(unique(GOI_cBP_mutations$altered_case_id))`**  
Total number of unique variants for **`r GOI`** returned:  **`r length(unique(GOI_cBP_mutations$amino_acid_change))`**  

## gnomAD Query Metadata

Total number of variants returned by gnomAD (non-cancer-associated and filtered): **`r length(GOI_gnomAD_df_filtered[,1])`**  
Number of unique variants found in both cBioPortal and gnomAD data: **`r sum(GOI_cBP_mutations$cBPgnomAD_overlap)`**

```{r echo=FALSE}
kable(multi_mut_table, row.names = FALSE, align = 'l',caption = "Cases with multiple variants")
```

```{r echo=FALSE}
kable(all_tissue_types_table, row.names = FALSE, align = 'l',caption = "Tissue types in all cases") #fix duplicates
```

# Protein features: Domains, Regions, Motifs and Modifications

```{r echo=FALSE, fig.height=5, fig.width=10, message=FALSE, warning=FALSE, out.width='\\textwidth', strip.white=TRUE}
if(!skip_F1){
  F1_magick_img <- image_graph(width = 1500, height = 1000,res = 150, clip = FALSE) #Graphics device that produces a Magick image
  lolliplot(SNP.gr = F1_variants,
            features = F1_features,
            ranges = F1_ranges,
            ylab = FALSE,
            xaxis = F1_x_axis)
  invisible(dev.off()) # shuts down graphics device
  par(mar=c(0,0,0,0)) # (bottom,left,top,right)
  plot(image_trim(F1_magick_img))
}else print("NO FEATURES RETURNED BY UNIPROT")
```


```{r echo=FALSE}
kable(GOI_protein_feature_annotation, row.names = FALSE, caption = "All features (minus variants and sequence conflicts)")
```


# Filtered Ensembl Transcripts (Union of Exonic sequence)

All transcripts are ploted to their relative genomic positions with all shared intronic regions removed. Transcription start sites are always on the left regardless of coding strand. **'Main' representative transcript used for variant alignment is in red**  
```{r echo=FALSE,fig.width=14,fig.height=10,out.width='\\textwidth',strip.white=TRUE}
F2_magick_img <- image_graph(width = 2000, height = 2000,res = 150, clip = FALSE)
lolliplot(SNP.gr = F2_variants,
             features = F2_features,
             ylab.gp = gpar(cex=1.25,fontface="bold"),
             ylab = "Filtered transcripts [Union of exonic sequence]\n\n",
             xaxis = F2_x_axis,
             yaxis = FALSE)
invisible(dev.off())
par(mar=c(0,0,0,0)) # (bottom,left,top,right)
plot(image_trim(F2_magick_img))
```

'Main' transcript: **`r GOI_TRANSCRIPT`**  
[Link to transcript support descriptions](https://useast.ensembl.org/info/genome/genebuild/transcript_quality_tags.html)  

```{r echo=FALSE}
kable(GOI_transcript_support[,2:5], row.names = FALSE, caption = "Filtered transcripts returned by ensemble with support annotations")
```

# All cBioPortal variants returned (n = **`r length(F3_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,fig.fullwidth = TRUE}
if(length(F3_variants$score)>0){
  F3_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F3_variants,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = legend
  )
  invisible(dev.off())
  F3_magick_img2 <- image_graph(width = image_info(image_trim(F3_magick_img))$width,
                                height = image_info(image_trim(F3_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,0,0)) # (bottom,left,top,right)
  plot(image_trim(F3_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F3_magick_img2)
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Missense mutations (n = **`r length(F4_variants$score)`**) [colors categorize reference AA]

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F4_variants$score)>0){
  F4_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F4_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = amino_acid_legend
  )
  invisible(dev.off())
  F4_magick_img2 <- image_graph(width = image_info(image_trim(F4_magick_img))$width,
                                height = image_info(image_trim(F4_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,0,0)) # (bottom,left,top,right)
  plot(image_trim(F4_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  #add custom legend for shapes
  legend("top",
         c("Gain of charge","Loss of charge","Charge reversed","Multi-category","No charge change"),
         pch = c(2,6,5,0,1), # "triangle point up","triangle point down","diamond","square"
         cex = 1.5,
         bty = "n",
         ncol = 5,
         inset = 0.08
         )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F4_magick_img2)
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Frameshifts, stops gained/lost and start lost (n = **`r length(F5_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F5_variants$score)>0){
  F5_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F5_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F5_types],
                          fill = legend$fill[legend$labels %in% F5_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  F5_magick_img2 <- image_graph(width = image_info(image_trim(F5_magick_img))$width,
                                height = image_info(image_trim(F5_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,0,0)) # (bottom,left,top,right)
  plot(image_trim(F5_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F5_magick_img2)
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Insertions/Deletions  (n = **`r length(F6_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F6_variants$score)>0){
  F6_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F6_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F6_types],
                          fill = legend$fill[legend$labels %in% F6_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  F6_magick_img2 <- image_graph(width = image_info(image_trim(F6_magick_img))$width,
                                height = image_info(image_trim(F6_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,0,0)) # (bottom,left,top,right)
  plot(image_trim(F6_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F6_magick_img2)
} else cat("\tNone returned by query\n\n")
```

### cBioPortal Splice regions and splice sites (n = **`r length(F7_variants$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F7_variants$score)>0){
  F7_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F7_variants ,
            features = c(F7_features,F3_features), 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F7_types],
                          fill = legend$fill[legend$labels %in% F7_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  F7_magick_img2 <- image_graph(width = image_info(image_trim(F7_magick_img))$width,
                                height = image_info(image_trim(F7_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,1,0)) # (bottom,left,top,right)
  plot(image_trim(F7_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  grid.text(paste0("[ Main transcript exons in green and red as top feature row ]"),
            x=0.04,
            y=0.95,
            just="left",
            gp=gpar(cex=1.25, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F7_magick_img2)
} else cat("\tNone returned by query\n\n")

```
```{r echo=FALSE}
kable(F7_exons, row.names = FALSE, caption = "Main transcript exons key", align = "l")
```

## gnomAD Splice regions and splice sites (n = **`r length(F7_variants$score)`**) in context of all transcripts [union of exonic sequence]

```{r echo=FALSE,fig.width=12,fig.height=12,out.width='\\textwidth',strip.white=TRUE}
if(length(F7_variants$score)>0){
  F8_magick_img <- image_graph(width = 2000, height = 2000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F8_variants,
               features = F8_features,
               ylab.gp = gpar(cex=1.25,fontface="bold"),
               ylab = "Filtered transcripts [Union of exonic sequence]",
               xaxis = F2_x_axis,
               yaxis.gp = gpar(cex = 1.5),
               xaxis.gp = gpar(cex = 1.5),
               cex = 1,
               jitter = "label",
               legend = list(labels = legend$labels[legend$labels %in% F7_types],
                            fill = legend$fill[legend$labels %in% F7_types],
                            cex = 1.5
                            )
               )
  invisible(dev.off())
  F8_magick_img2 <- image_graph(width = image_info(image_trim(F8_magick_img))$width,
                                height = image_info(image_trim(F8_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,1.5,1,0)) # (bottom,left,top,right)
  plot(image_trim(F8_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  grid.text(paste0("[ Main transcript exons in green\nand red as top feature row ]"),
            x=0.04,
            y=0.90,
            just="left",
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F8_magick_img2)
} else cat("\tNone returned by query\n\n")
```


# All gnomAD variants returned (n = **`r length(F9_gnomAD$score)`**)

```{r echo=FALSE, fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F9_gnomAD$score)>0){
  F9_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F9_gnomAD,
            features = F9_features, 
            ranges = F9_ranges, 
            xaxis = F9_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  legend)
  invisible(dev.off())
  F9_magick_img2 <- image_graph(width = image_info(image_trim(F9_magick_img))$width,
                                height = image_info(image_trim(F9_magick_img))$height,
                                res = 150,
                                clip = FALSE
  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F9_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F9_magick_img2)
} else cat("\tNone returned by query\n\n")
```

## gnomAD Missense mutations (n = **`r length(F10_gnomAD$score)`**) [colors categorize reference AA]

```{r echo=FALSE, fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F10_gnomAD$score)>0){
  F10_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F10_gnomAD,
            features = F9_features, 
            ranges = F9_ranges, 
            xaxis = F9_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  amino_acid_legend)
  invisible(dev.off())
  F10_magick_img2 <- image_graph(width = image_info(image_trim(F10_magick_img))$width,
                                height = image_info(image_trim(F10_magick_img))$height,
                                res = 150,
                                clip = FALSE
  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F10_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",F10_max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  #add custom legend for shapes
  legend("top",
         c("Gain of charge","Loss of charge","Charge reversed","Multi-category","No charge change"),
         pch = c(2,6,5,0,1), # "triangle point up","triangle point down","diamond","square"
         cex = 1.5,
         bty = "n",
         ncol = 5,
         inset = 0.08
         )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F10_magick_img2)
} else cat("\tNone returned by query\n\n")
```

## gnomAD Frameshifts and stops gained/lost (n = **`r length(F11_gnomAD$score)`**)

```{r echo=FALSE, fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F11_gnomAD$score)>0){
  F11_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F11_gnomAD,
            features = F9_features, 
            ranges = F9_ranges, 
            xaxis = F9_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  list(labels = legend$labels[legend$labels %in% F11_types],
                          fill = legend$fill[legend$labels %in% F11_types],
                          cex = 1.5
                          )
            )
  invisible(dev.off())
  F11_magick_img2 <- image_graph(width = image_info(image_trim(F11_magick_img))$width,
                                height = image_info(image_trim(F11_magick_img))$height,
                                res = 150,
                                clip = FALSE
  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F11_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",F11_max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F11_magick_img2)
} else cat("\tNone returned by query\n\n")
```

## gnomAD Insertions/Deletions  (n = **`r length(F12_gnomAD$score)`**)

```{r echo=FALSE, fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F12_gnomAD$score)>0){
  F12_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F12_gnomAD,
            features = F9_features, 
            ranges = F9_ranges, 
            xaxis = F9_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  list(labels = legend$labels[legend$labels %in% F12_types],
                          fill = legend$fill[legend$labels %in% F12_types],
                          cex = 1.5
                          )
            )
  invisible(dev.off())
  F12_magick_img2 <- image_graph(width = image_info(image_trim(F12_magick_img))$width,
                                height = image_info(image_trim(F12_magick_img))$height,
                                res = 150,
                                clip = FALSE
  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F12_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",F12_max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F12_magick_img2)
} else cat("\tNone returned by query\n\n")
```

## gnomAD Splice regions and splice sites (n = **`r length(F13_gnomAD$score)`**)

```{r echo=FALSE,fig.width=14,fig.height=4.5,out.width='\\textwidth'}
if(length(F13_gnomAD$score)>0){
  F13_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F13_gnomAD ,
            features = c(F7_features,F9_features), 
            ranges = F9_ranges, 
            xaxis = F9_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = list(labels = legend$labels[legend$labels %in% F13_types],
                          fill = legend$fill[legend$labels %in% F13_types],
                          cex = 1.5
                          )
  )
  invisible(dev.off())
  F13_magick_img2 <- image_graph(width = image_info(image_trim(F13_magick_img))$width,
                                height = image_info(image_trim(F13_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3.5,1,0)) # (bottom,left,top,right)
  plot(image_trim(F13_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",F13_max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  grid.text(paste0("[ Main transcript exons in green and red as top feature row ]"),
            x=0.04,
            y=0.95,
            just="left",
            gp=gpar(cex=1.25, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F13_magick_img2)
} else cat("\tNone returned by query\n\n")

```

## gnomAD Splice regions and splice sites (n = **`r length(F13_gnomAD$score)`**) in context of all transcripts [union of exonic sequence]

```{r echo=FALSE,fig.width=12,fig.height=12,out.width='\\textwidth',strip.white=TRUE}
if(length(F13_gnomAD$score)>0){
  F14_magick_img <- image_graph(width = 2000, height = 2000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F14_variants,
               features = F14_features,
               ylab = FALSE,
               xaxis = F2_x_axis,
               yaxis.gp = gpar(cex = 1.5),
               xaxis.gp = gpar(cex = 1.5),
               cex = 1,
               jitter = "label",
               legend = list(labels = legend$labels[legend$labels %in% F13_types],
                            fill = legend$fill[legend$labels %in% F13_types],
                            cex = 1.5
                            )
               )
  invisible(dev.off())
  F14_magick_img2 <- image_graph(width = image_info(image_trim(F14_magick_img))$width,
                                height = image_info(image_trim(F14_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3.5,1,0)) # (bottom,left,top,right)
  plot(image_trim(F14_magick_img))
  # yaxis label
  grid.text(paste0("log2(allele count)\n[max count = ",F13_max_gnomAD_allele_count,"]"),
            x=0.001,
            y=0.5,
            just="top",
            rot=90,
            gp=gpar(cex=1.5, fontface="bold")
            )
  grid.text(paste0("[ Main transcript exons in green\nand red as top feature row ]"),
            x=0.04,
            y=0.90,
            just="left",
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F14_magick_img2)
} else cat("\tNone returned by query\n\n")
```

# Overlapping Variants between cBioPortal and gnomAD (n = **`r length(unique_cBP_overlap_df[,1])`**)

```{r echo=FALSE, fig.height=8, fig.width=14,out.width='\\textwidth'}
if(length(unique_cBP_overlap_df[,1]) > 1){
  #legend
  F15_magick_img_legend <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  plot.new()
  legend("center",
         c("Does not meet filter threshold (gnomAD Allele Freq < 0.0001)","Meets filter threshold (gnomAD Allele Freq >= 0.0001)"),
         pch = c(1,0), # "triangle point up","triangle point down","diamond","square"
         cex = 1,
         #bty = "n",
         ncol = 2,
         inset = -0.1
         )
  invisible(dev.off())
  # top: cBP overlap
  F15_magick_img_top <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F15_cBP_overlap,
            features = F15_features,
            ranges = F3_ranges,
            xaxis = FALSE,
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  FALSE)
  invisible(dev.off())
  F15_magick_img_top2 <- image_graph(width = image_info(image_trim(F15_magick_img_top))$width,
                                  height = image_info(image_trim(F15_magick_img_top))$height,
                                  res = 150,
                                  clip = FALSE
                                  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,top,right)
  plot(image_trim(F15_magick_img_top))
  # yaxis label
  grid.text(paste0("cBioPortal: Variant count"),
              x=0.001,
              y=0.5,
              just="top",
              rot=90,
              gp=gpar(cex=1, fontface="bold")
              )
  invisible(dev.off())
  
  # bottom: gnomAD overlap
  F15_magick_img_bottom <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F15_gnomAD_overlap,
            features = F15_features,
            ranges = F3_ranges,
            xaxis = FALSE,
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend =  FALSE)
  invisible(dev.off())
  F15_magick_img_bottom2 <- image_graph(width = image_info(image_trim(F15_magick_img_bottom))$width,
                                  height = image_info(image_trim(F15_magick_img_bottom))$height,
                                  res = 150,
                                  clip = FALSE
                                  )
  par(mar=c(0,3.5,0,0)) # (bottom,left,bottom,right)
  plot(image_trim(F15_magick_img_bottom))
  # yaxis label
    grid.text(paste0("gnomAD: log2(allele count)\n[max count = ",F15_max_gnomAD_allele_count,"]"),
              x=0.001,
              y=0.5,
              just="top",
              rot=90,
              gp=gpar(cex=1, fontface="bold")
              )
  invisible(dev.off())
  
  par(mfrow=c(3,1))
  par(mar=c(0,0,0,0))
  plot(image_trim(F15_magick_img_legend))
  plot(F15_magick_img_top2)
  plot(F15_magick_img_bottom2)
  
} else cat("\tNo overlapping variants\n\n")
```

```{r echo=FALSE}
kable(overlap_summary_table, row.names = FALSE, caption = "All variants shared between cBP and gnomAD", align = "l")
```

# Cancer-specific hits after filtering against gnomAD natural variants

Significance is assesed via an exact binomial test:  
After filtering against natural variation, the number of trials equals the total number of variants observed, the probability of success equals 1/(number of unique variants) and trials equals the total number of variants.  
P-value <= 0.01 is considered significant
This test makes many assumtions (ignores sequence, codon, mutation burdon and structural contexts) and should only be considered as a crude measure of identifying the most significant peaks.  

```{r echo=FALSE,fig.width=14,fig.height=4.5,fig.fullwidth = TRUE}
if(length(F3_variants$score)>0){
  F16_magick_img <- image_graph(width = 3000, height = 3000,res = 150, clip = FALSE)
  lolliplot(SNP.gr = F16_variants ,
            features = F3_features, 
            ranges = F3_ranges, 
            xaxis = F3_x_axis,
            yaxis.gp = gpar(cex = 1.5),
            xaxis.gp = gpar(cex = 1.5),
            cex = 1,
            ylab = FALSE,
            jitter = "label",
            legend = legend
  )
  invisible(dev.off())
  F16_magick_img2 <- image_graph(width = image_info(image_trim(F16_magick_img))$width,
                                height = image_info(image_trim(F16_magick_img))$height,
                                res = 150,
                                clip = FALSE
                                )
  par(mar=c(0,3,0,0)) # (bottom,left,top,right)
  plot(image_trim(F16_magick_img))
  # yaxis label
  grid.text(paste0("Variant count"),
            x=0.01,
            y=0.6,
            just="top",
            rot=90,
            gp=gpar(cex=1, fontface="bold")
            )
  invisible(dev.off())
  par(mar=c(0,0,0,0))
  plot(F16_magick_img2)
} else cat("\tNone returned by query\n\n")
```

## All variants p-value <= 0.05 (significance threshold set @ 0.01)

```{r echo=FALSE}
if(length(hits_table$`Binomial test p-val` <= 0.01) >1){
  kable(hits_table[hits_table$`Binomial test p-val` <= 0.05,],
        row.names = FALSE
        )
}else{
  cat("\tNone returned by query\n\n")
}
```

# Gene fusions returned:

```{r echo=FALSE}
if(length(GOI_cBP_fusions[,1])>0){
  kable(GOI_cBP_fusions[,c("case_id","amino_acid_change","genetic_profile_id")],
        col.names = c("case_id","fusion",'genetic_profile_id'),
        row.names = FALSE
        )
}else{
  cat("\tNone returned by query\n\n")
}
```
